docs: add a how-to guide focused on debugging charms (#2386)
canonical/operator: docs/howto/debug-your-charm.md

```shell
juju debug-code --at=config-start myapp/0 config-changed
```

Python's built-in {external+python:func}`breakpoint` also works when `JUJU_DEBUG_AT` is set, so you can use either form.

> See more: {external+juju:ref}`Juju | juju debug-code <command-juju-debug-code>`

(remote-debugging-with-vs-code)=
## Remote debugging with VS Code

For a richer debugging experience, you can attach VS Code's debugger to a running charm using [`debugpy`](https://github.com/microsoft/debugpy). This gives you a full graphical debugger with breakpoints, variable inspection, watch expressions, and call stack navigation.

### Set up the charm

Add `debugpy` as a dependency of your charm. For example, with uv run `uv add debugpy` to add it to the `[project]` dependencies in `pyproject.toml`; with Poetry run `poetry add debugpy`; or with a plain `requirements.txt`, add a `debugpy` line.

Then, in your charm code, add a `debugpy` listener that activates when `JUJU_DEBUG_AT` is set (i.e. when you run `juju debug-code`):

```python
import os

if os.getenv('JUJU_DEBUG_AT'):
    import debugpy
    debugpy.listen(('0.0.0.0', 5678))
    debugpy.wait_for_client()

> ```

Reviewer: I wonder if it would be cleaner to provide a recipe that uses the built-in `breakpoint()` to trigger `debugpy`. Ops overrides `sys.breakpointhook` to point to `Framework.breakpoint`, so we'd get the correct `JUJU_DEBUG_AT` semantics for free.

The debugpy readme gives [this example](https://github.com/microsoft/debugpy?tab=readme-ov-file#breakpoint-function) which looks promising -- `listen` at the top of `charm.py` and then just use `breakpoint()` as usual might 'just work'.
```py
import debugpy
debugpy.listen(...)

while True:
    ...
    breakpoint()  # or debugpy.breakpoint()
    ...
```

Author: As per other thread, calling `breakpoint()` in the code is bad: it drops you into `pdb` in the place you ran `debug-code`, and there's a weird message about that in VS Code rather than actually connecting to the debugging functionality in VS Code.

Place this at the top of the event handler you want to debug, or in your charm's `__init__` to intercept every hook.

Repack and deploy the charm with `charmcraft pack`.

### Configure VS Code

Add the following launch configuration to `.vscode/launch.json` in your charm project:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Attach to charm",
            "type": "python",
            "request": "attach",
            "connect": {
                "host": "<UNIT_IP>",
                "port": 5678
            },
            "pathMappings": [
                {
                    "localRoot": "${workspaceFolder}",
                    "remoteRoot": "."
                }
            ],
            "justMyCode": true
        }
    ]
}
```

Find the unit IP with:

```shell
juju show-unit myapp/0 | yq '.*.address'
```

### Start a debug session

1. Run `juju debug-code myapp/0` to tell Juju to set `JUJU_DEBUG_AT` on the next hook execution.
2. Trigger the hook you want to debug (or wait for it to fire naturally). The charm will start `debugpy` and block until a client connects.

> 3. In VS Code, set your breakpoints and press **F5** (or click **Run > Start Debugging**).

Author: Yes, if you're using the VS Code approach, then you use the breakpoints in VS Code, you don't put `breakpoint()` or `framework.breakpoint()` calls in the code. I don't know what's going on under the hood, but that's the exact case that I validated works.

And, I believe, you *shouldn't* put breakpoints into the code, because those will try to go out to a tmux session that you're not using.

(debug-your-charm)=
# How to debug your charm

> See first: {ref}`log-from-your-charm`, {external+juju:ref}`Juju | How to manage logs <manage-logs>`

When your charm isn't behaving as expected, Juju and the broader charming ecosystem provide several tools to help you investigate. This guide covers how to get a shell on a running unit, read charm logs, interactively debug hook execution, and use third-party tools to accelerate your debugging workflow.

> 

Reviewer 2: How about adding a link to the 12-factor advice, which is pretty comprehensive:

```suggestion

For advice about debugging a 12-factor app charm, also see {external+charmcraft:ref}`Charmcraft | Use a 12-factor app charm <use-12-factor-charms-troubleshoot>`.

```

(ssh-into-a-unit)=
## Get a shell on a running unit with `juju ssh`

The quickest way to poke around a live unit is to SSH into it with `juju ssh`. What you connect to depends on the substrate:

**Machine charms.** `juju ssh` connects you to the machine itself. You land in a shell as the `ubuntu` user, which has password-less `sudo`:

```shell
juju ssh myapp/0

> ```

Reviewer 2: Very minor - I'd put the "You land in" part after the command. Otherwise it looks like we're introducing a command related to the `ubuntu` user.

````suggestion
**Machine charms.** `juju ssh` connects you to the machine itself:

```shell
juju ssh myapp/0
```

You land in a shell as the `ubuntu` user, which has password-less `sudo`.
````

**Kubernetes charms.** A K8s charm pod typically has multiple containers (the charm container and one or more workload containers). Use `--container` to choose which container to connect to:

```shell
juju ssh --container myworkload myapp/0
```

If you omit `--container`, `juju ssh` targets the charm container by default.

````{important}
**Juju 4: SSH keys are no longer added automatically.**

In Juju 3.x, your SSH public key is automatically added to every model you create. In Juju 4, this no longer happens -- you must add your key explicitly before `juju ssh` will work:

```shell
juju add-ssh-key "$(cat ~/.ssh/id_ed25519.pub)"
```

You can also import keys from GitHub or Launchpad:

```shell
juju import-ssh-key gh:<your-github-username>
```
````

> See more: {external+juju:ref}`Juju | juju ssh <command-juju-ssh>`

(debug-with-pebble)=
## Inspect the workload with Pebble (Kubernetes charms)

> In Kubernetes charms, each workload container runs {external+pebble:doc}`Pebble <index>` as its init system. You can use Pebble commands to inspect and interact with the workload directly. First, SSH into the workload container:

Reviewer 2: I haven't seen us use "init system" before. How about "service manager" instead?

```suggestion
In Kubernetes charms, each workload container runs {external+pebble:doc}`Pebble <index>` as its service manager. You can use Pebble commands to inspect and interact with the workload directly. First, SSH into the workload container:
```

```shell
sudo snap install jhack
sudo snap connect jhack:dot-local-share-juju snapd
```

### Monitor events with `jhack tail`

`jhack tail` watches the Juju log and displays charm events in a colour-coded, formatted table. It is much easier to scan than raw `juju debug-log` output when you want to understand the flow of events:

```shell
jhack tail myapp
```

### Trigger events with `jhack fire`

`jhack fire` simulates a specific event on a live unit. This is useful for triggering an event on demand without waiting for Juju to emit it naturally:

```shell
jhack fire myapp/0 update-status
jhack fire myapp/0 config-changed
```

```{caution}
Firing events manually can desynchronise charm state from Juju state if your event handlers are not idempotent. Use this only in development and test environments.
```

### Push local changes with `jhack sync`

`jhack sync` watches local directories and automatically pushes file changes to remote charm units. Combined with `jhack fire`, this enables a rapid edit-trigger-observe loop:

```shell
jhack sync myapp/0 --source ./src --source ./lib

> ```

Reviewer 2: Similar suggestion as in a couple of other places

````suggestion
`jhack sync` watches local directories and automatically pushes file changes to remote charm units:

```shell
jhack sync myapp/0 --source ./src --source ./lib
```

Combined with `jhack fire`, this enables a rapid edit-trigger-observe loop.
````

```shell
juju debug-hooks myapp/0                       # intercept all hooks and actions
juju debug-hooks myapp/0 config-changed        # intercept only config-changed
```

*Once a hook fires*, the `tmux` session lands in the charm directory. From there you can:

- Inspect the environment variables that Juju provides (e.g. `JUJU_DISPATCH_PATH`).
- Examine or modify files under `src/`.
- Run `./dispatch` to execute the hook manually.
- Run `./dispatch` again after making changes, to iterate.
- Exit the `tmux` session to let the unit resume normal operation.

```{note}
While a hook is being debugged, the unit is paused. Other hooks queue up and execute in order once you exit. Keep your debugging sessions short to avoid blocking the unit for too long.
```

> See more: {external+juju:ref}`Juju | juju debug-hooks <command-juju-debug-hooks>`

(debug-code)=
## Step through charm code with `juju debug-code`

> The `juju debug-code` command is similar to `debug-hooks`, but the hook **is** executed automatically. Juju sets the `JUJU_DEBUG_AT` environment variable, which Ops uses to activate breakpoints. When execution reaches a breakpoint, you are dropped into a {external+python:mod}`pdb` session where you can inspect variables and step through the code.

Reviewer 2: ```suggestion
The `juju debug-code` command is similar to `debug-hooks`, but the hook *is* executed automatically. Juju sets the `JUJU_DEBUG_AT` environment variable, which Ops uses to activate breakpoints. When execution reaches a breakpoint, you are dropped into a {external+python:mod}`pdb` session where you can inspect variables and step through the code.
```
